src/plugins/vis_augmenter/public/vega/helpers.ts

```diff
@@ -0,0 +1,324 @@
+
+  if (!isEmpty(augmentedTable.rows)) {
+    const dataStartTime = augmentedTable.rows[0][xAxisId] as number;
+    const dataEndTime = augmentedTable.rows[augmentedTable.rows.length - 1][xAxisId] as number;
+
+    let curStartTime = dataStartTime;
+    while (curStartTime > chartStartTime) {
+      curStartTime -= intervalMillis;
+      augmentedTable.rows.unshift({
+        [xAxisId]: curStartTime,
+      });
+    }
+
+    let curEndTime = dataEndTime;
+    while (curEndTime < chartEndTime) {
+      curEndTime += intervalMillis;
+      augmentedTable.rows.push({
+        [xAxisId]: curEndTime,
+      });
+    }
+  } else {
+    // if there's no existing rows, create them all
+    let curTime = chartStartTime;
+    while (curTime <= chartEndTime) {
+      augmentedTable.rows.push({
+        [xAxisId]: curTime,
+      });
+      curTime += intervalMillis;
+    }
+  }
+  return augmentedTable;
+};
+
+/**
+ * Adding events into the correct x-axis key (the time bucket)
+ * based on the table. As of now only results from
+ * PointInTimeEventsVisLayers are supported
+ */
+export const addPointInTimeEventsLayersToTable = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any,
+  visLayers: PointInTimeEventsVisLayer[]
+): OpenSearchDashboardsDatatable => {
+  const augmentedTable = addMissingRowsToTableBounds(datatable, dimensions);
+  const xAxisId = getXAxisId(dimensions, augmentedTable.columns);
+
+  if (!isEmpty(visLayers)) {
+    visLayers.every((visLayer: PointInTimeEventsVisLayer) => {
+      const visLayerColumnId = `${visLayer.pluginResource.id}`;
+      const visLayerColumnName = `${visLayer.pluginResource.name}`;
+      augmentedTable.columns.push({
+        id: visLayerColumnId,
+        name: visLayerColumnName,
+        meta: {
+          type: VIS_LAYER_COLUMN_TYPE,
+        },
+      });
+
+      if (augmentedTable.rows.length === 0) {
+        return false;
+      }
+
+      // if only one row / one datapoint, put all events into this bucket
+      if (augmentedTable.rows.length === 1) {
+        augmentedTable.rows[0] = {
+          ...augmentedTable.rows[0],
+          [visLayerColumnId]: visLayer.events.length,
+        };
+        return false;
+      }
+
+      // Bin the timestamps to the closest x-axis key, adding
+      // an entry for this vis layer ID. Sorting the timestamps first
+      // so that we will only search a particular row value once.
+      // There could be some optimizations, such as binary search + dynamically
+      // changing the bounds, but performance benefits would be very minimal
+      // if any, given the upper bounds limit on n already due to chart constraints.
+      let rowIndex = 0;
+      const sortedTimestamps = visLayer.events
+        .map((event: PointInTimeEvent) => event.timestamp)
+        .sort((n1: number, n2: number) => n1 - n2) as number[];
+
+      if (sortedTimestamps.length > 0) {
+        sortedTimestamps.forEach((timestamp) => {
```

Comment:
Shouldnt the forEach handle the empty `sortedTimestamps` case, so there wont be a need to have the if statement.

Reply:
Yes thanks, I'll remove